cli/best_ancestor.go

```diff
@@ -118,34 +103,92 @@ func fastFindAncestor(r *git.Repository) (h plumbing.Hash, c int, branchName str
 	return
 }
 
-func findRemoteBranch(reference *plumbing.Reference, r *git.Repository) (remoteBranch *plumbing.Reference, err error) {
+func (p *RemoteCmd) findRemoteBranch(ctx context.Context, reference *plumbing.Reference, r *git.Repository) (remoteBranch *plumbing.Reference, err error) {
 	branchIter, err := r.References()
 	if err != nil {
 		return
 	}
 	// If branch is main, just point to it directly
-	remoteDefault := "refs/remote/origin/main"
+	remoteDefault := "refs/remotes/origin/main"
 	remoteBranch, err = r.Reference(plumbing.ReferenceName(remoteDefault), false)
 	if err == nil {
 		return
 	}
 	// If branch is master, just point to it directly
-	remoteDefault = "refs/remote/origin/master"
+	remoteDefault = "refs/remotes/origin/master"
 	remoteBranch, err = r.Reference(plumbing.ReferenceName(remoteDefault), false)
 	if err == nil {
 		return
 	}
 	// We have some issues here
-	// TODO(ch2036): Maybe switch to using `git` utility directly
-	err = branchIter.ForEach(func(rem *plumbing.Reference) error {
-		if rem.Name().IsRemote() {
-			log.Debugf("Branch found: %v, %s ?== %s", rem, rem.Name().Short(), reference.Name().Short())
-			if strings.HasSuffix(rem.Name().Short(), reference.Name().Short()) {
-				remoteBranch = rem
-				return storer.ErrStop
+	if p.goGit {
+		log.Warnf("go-git traverse of references can lead to issues...")
+		err = branchIter.ForEach(func(rem *plumbing.Reference) error {
+			if rem.Name().IsRemote() {
+				log.Debugf("Branch found: %v, %s ?== %s", rem, rem.Name().Short(), reference.Name().Short())
+				if strings.HasSuffix(rem.Name().Short(), reference.Name().Short()) {
+					remoteBranch = rem
+					return storer.ErrStop
+				}
+			}
+			return nil
+		})
+	} else {
+		const (
+			gitFetchCmd  = " fetch --all"
+			gitBranchCmd = " branch --remote"
+		)
+
+		cmdString := gitExe + gitFetchCmd
+		buf := bytes.NewBuffer(nil)
+		log.Debugf("Executing '%v'...", cmdString)
+		if err = p.metalTgt.ExecSilentlyToWriter(ctx, cmdString, p.repoDir, buf); err != nil {
+			err = fmt.Errorf("git fetch: %w", err)
+			return
+		}
+
+		if buf.Len() == 0 {
+			err = fmt.Errorf("git fetch: empty response")
+			return
+		}
+		branchFetched := strings.Count(buf.String(), "\n")
+		log.Infof("Feched %d branches", branchFetched)
+		buf.Truncate(0)
+
+		// Now check all remote branches names
+		cmdString = gitExe + gitBranchCmd
+		log.Debugf("Executing '%v'...", cmdString)
+		if err = p.metalTgt.ExecSilentlyToWriter(ctx, cmdString, p.repoDir, buf); err != nil {
+			err = fmt.Errorf("git branch: %w", err)
+			return
+		}
+
+		if buf.Len() == 0 {
+			err = fmt.Errorf("git branch: empty response")
+			return
+		}
+		for {
+			line, rerr := buf.ReadString('\n')
+			if rerr != nil {
+				if errors.Is(rerr, io.EOF) {
+					err = fmt.Errorf("reached %w", rerr)
+				} else {
+					err = rerr
+				}
+				return
+			}
+			line = strings.Trim(line, " \n")
+			bName := strings.ReplaceAll(line, "origin/", "")
```

Comment:
Depending on a particular remote name can be quite brittle. Is there a way we can get around this?